### PR TITLE
Implement backcompat for Futures.addCallback + Futures.addFallback

### PR DESCRIPTION
**Pull request by carterkozak**
==COMMIT_MSG==
Implement backcompat for Futures.addCallback + Futures.addFallback
==COMMIT_MSG==

### CODE REVIEW

File: guava-compatibility-agent/build.gradle [hunk 1]
```diff
@@ -3,6 +3,7 @@ apply plugin: 'com.palantir.shadow-jar'
 
 dependencies {
     shadeTransitively 'net.bytebuddy:byte-buddy'
+    compileOnly 'com.google.guava:guava:32.1.2-jre'
```
Comment (bjlaub) [inline guava-compatibility-agent/build.gradle:6]: out of curiosity, what's the reason we only needed to add this now? I guess previously this code only accessed guava types via bytebuddy at runtime?

**Review by bjlaub — APPROVED** [review]

File: guava-compatibility-agent/src/main/java/com/palantir/guavacompat/agent/Agent.java [hunk 2]
```diff
@@ -74,10 +76,12 @@ public static void premain(String _args, Instrumentation instrumentation) {
         agentBuilder = agentBuilder
                 .type(ElementMatchers.named(FUTURES))
                 .transform((builder, _type, _classLoader, _module, _protection) -> {
-                    TypeDescription.Generic inputTypeVariable =
+                    TypeDescription.Generic typeVarI =
                             TypeDescription.Generic.Builder.typeVariable("I").build();
-                    TypeDescription.Generic outputTypeVariable =
+                    TypeDescription.Generic typeVarO =
                             TypeDescription.Generic.Builder.typeVariable("O").build();
+                    TypeDescription.Generic typeVarV =
```
Comment (mpritham) [inline guava-compatibility-agent/src/main/java/com/palantir/guavacompat/agent/Agent.java:83]: ~Should we use this in this code below?:~

```@Override
public TypeList.Generic getTypeVariables() {
  return new TypeList.Generic.Explicit(
    TypeDescription.Generic.Builder.typeVariable("V")
      .build());
}
```

edit: on second thought, I don't think so. The `getTypeVariables` methods are used when defining a different type.

File: guava-compatibility-agent/src/main/java/com/palantir/guavacompat/agent/FallbackAdapterHandler.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import java.util.concurrent.Executor;
+
+public final class FallbackAdapterHandler {
+    public static <V> ListenableFuture<V> withFallback(
+            ListenableFuture<? extends V> input, FutureFallback<? extends V> fallback) {
+        return Futures.catchingAsync(
+                input, Throwable.class, new FallbackAdapter<>(fallback), MoreExecutors.directExecutor());
```
Comment (mpritham) [inline guava-compatibility-agent/src/main/java/com/palantir/guavacompat/agent/FallbackAdapterHandler.java:29]: naive q: can we pass in `fallback` directly? why do we need the `FallbackAdapter` here?

**Review by mpritham — APPROVED** [review]

File: guava-compatibility-agent/build.gradle (hunk 1 above)
Comment (carterkozak) [reply guava-compatibility-agent/build.gradle:6]: This compile-only dependency is used to implement `FallbackAdapterHandler` which delegates to `Futures.catchingAsync` as well as `FallbackAdapter` which implements `AsyncFunction<Throwable, T>`.

That could be done using byte-buddy to define classes, but it's a fair bit trickier.

File: guava-compatibility-agent/src/main/java/com/palantir/guavacompat/agent/Agent.java (hunk 2 above)
Comment (carterkozak) [reply guava-compatibility-agent/src/main/java/com/palantir/guavacompat/agent/Agent.java:83]: I think you're right, good call -- The `TypeList` can wrap the typeVariables we've defined above -- I'll fix it up!

**Pull request merged**